### PR TITLE
Upgrade torchmetrics

### PR DESCRIPTION
**Pull request by nik-mosaic**
# What does this PR do?

Upgrades torchmetrics to support 0.11.x. This is not straightforward because torchmetrics.Accuracy() now requires a task argument (either 'binary', 'multiclass', or 'multilabel'). This PR replaces all instances of torchmetrics.Accuracy() in Composer with multiclass Accuracy, with the correct number of classes provided. torchmetrics.MatthewsCorrCoef() is also updated similarly.

Because torchmetrics.Accuracy(task='multiclass') requires a num_classes argument, we add a new, optional 'num_classes' argument to ComposerClassifier. Users must either
(1) Pass in num_classes to the ComposerClassifier, or
(2) Specify a num_classes parameter in the PyTorch network submodule.

This is because if we are to use Accuracy as our default training and validation metric, we have to know how many classes the user wants. Alternatively, if users do not want to use our default training/validation metric, they can
(3) Pass in both train_metrics and val_metrics to Composer Classifier
and we will not require `num_classes` since we no longer need to create a default metric.

If none of these three options are satisfied, ComposerClassifier will now raise an error.

The rest of the changes in this PR are fixing the various models, tests, notebooks, and documentation snippets that depend on ComposerClassifier and properties of its Accuracy metric, such as the class name or shape.
 
# What issue(s) does this change relate to?
[CO-1836](https://mosaicml.atlassian.net/browse/CO-1836)

[CO-1836]: https://mosaicml.atlassian.net/browse/CO-1836?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/models/base.py [hunk 1]
```diff
@@ -184,13 +184,13 @@ def metrics(self, train: bool = False) -> Dict[str, Metric]:
 
         .. code-block:: python
 
-            from torchmetrics.classification import Accuracy
+            from torchmetrics.classification import MultitaskAccuracy
```
Comment (dakinggg) [inline composer/models/base.py:187]: Multiclass?

File: composer/models/base.py [hunk 2]
```diff
@@ -184,13 +184,13 @@ def metrics(self, train: bool = False) -> Dict[str, Metric]:
...
             from composer.models.loss import CrossEntropyLoss
 
-            def __init__(self):
+            def __init__(self, num_classes):
                 super().__init__()
-                self.train_acc = Accuracy() # torchmetric
-                self.val_acc = Accuracy()
+                self.train_acc = MultitaskAccuracy(num_classes=num_classes) # torchmetric
```
Comment (dakinggg) [inline composer/models/base.py:192]: Multiclass?

File: composer/models/base.py [hunk 3]
```diff
@@ -184,13 +184,13 @@ def metrics(self, train: bool = False) -> Dict[str, Metric]:
...
+                self.val_acc = MultitaskAccuracy(num_classes=num_classes)
```
Comment (dakinggg) [inline composer/models/base.py:193]: Multiclass?

File: composer/models/bert/model.py [hunk 4]
```diff
@@ -104,24 +103,24 @@ def create_bert_mlm(use_pretrained: Optional[bool] = False,
...
-                               gradient_checkpointing: Optional[bool] = False):
+                               gradient_checkpointing: bool = False):
     """BERT classification model based on |:hugging_face:| Transformers.
 
     For more information, see `Transformers <https://huggingface.co/transformers/>`_.
 
     Args:
         num_labels (int, optional): The number of classes in the classification task. Default: ``2``.
-        gradient_checkpointing (bool, optional): Use gradient checkpointing. Default: ``False``.
+            gradient_checkpointing (bool, optional): Use gradient checkpointing. Default: ``False``.
```
Comment (dakinggg) [inline composer/models/bert/model.py]: bad indent?

File: composer/models/tasks/classification.py [hunk 5]
```diff
@@ -49,34 +51,44 @@ class ComposerClassifier(ComposerModel):
...
+                 num_classes: Optional[int] = None,
                  train_metrics: Optional[Union[Metric, MetricCollection]] = None,
                  val_metrics: Optional[Union[Metric, MetricCollection]] = None,
                  loss_fn: Callable = soft_cross_entropy) -> None:
         super().__init__()
 
+        self.module = module
+        self._loss_fn = loss_fn
+
+        self.num_classes = num_classes
```
Comment (dakinggg) [inline composer/models/tasks/classification.py:70]: which should we take if `num_classes` is passed in and an attr on the module? should we warn that both were set and say which one we're using? could be some confusing errors other wise

File: composer/models/tasks/classification.py [hunk 6]
```diff
@@ -49,34 +51,44 @@ class ComposerClassifier(ComposerModel):
...
+        if hasattr(self.module, 'num_classes'):
+            self.num_classes = getattr(self.module, 'num_classes')
+        if self.num_classes == None and (train_metrics is None or val_metrics is None):
```
Comment (dakinggg) [inline composer/models/tasks/classification.py:73]:
```suggestion
        if self.num_classes is None and (train_metrics is None or val_metrics is None):
```

File: composer/models/tasks/classification.py [hunk 7]
```diff
@@ -49,34 +51,44 @@ class ComposerClassifier(ComposerModel):
...
+                textwrap.dedent('Please specify the number of output classes. Either: \n (1) pass '
+                                'in num_classes to the ComposerClassifier \n (2) pass in both '
+                                'train_metrics and val_metrics to Composer Classifier, or \n (3) '
+                                'specify a num_classes parameter in the PyTorch network module.'))
+
         # Metrics for training
         if train_metrics is None:
-            train_metrics = Accuracy()
+            assert self.num_classes is not None
+            train_metrics = MulticlassAccuracy(num_classes=self.num_classes)
```
Comment (dakinggg) [inline composer/models/tasks/classification.py:83]: just to double check, does `MulticlassAccuracy` just become `BinaryAccuracy` if you pass `num_classes=2`?

File: docs/source/composer_model.rst [hunk 8]
```diff
@@ -134,8 +134,8 @@ A full example of a validation implementation would be:
         def __init__(self):
             super().__init__()
             self.model = torchvision.models.resnet18()
-            self.train_accuracy = torchmetrics.Accuracy()
-            self.val_accuracy = torchmetrics.Accuracy()
+            self.train_accuracy = torchmetrics.classification.Accuracy(num_classes=1000)
```
Comment (dakinggg) [inline docs/source/composer_model.rst:137]: Multiclass?

File: docs/source/composer_model.rst [hunk 9]
```diff
@@ -134,8 +134,8 @@ A full example of a validation implementation would be:
...
+            self.val_accuracy = torchmetrics.classification.Accuracy(num_classes=1000)
```
Comment (dakinggg) [inline docs/source/composer_model.rst:138]: Multiclass?

File: setup.py [hunk 10]
```diff
@@ -75,7 +75,7 @@ def package_files(prefix: str, directory: str, extension: str):
 install_requires = [
     'pyyaml>=6.0,<7',
     'tqdm>=4.62.3,<5',
-    'torchmetrics>=0.7.0,<0.10.0',
+    'torchmetrics>=0.10.0,<0.12.0',
```
Comment (dakinggg) [inline setup.py:78]: for posterity's sake, could you prove that the code in this PR works with both 0.10 and 0.11 and put it in the PR description?

File: tests/algorithms/test_cutmix.py [hunk 11]
```diff
@@ -81,7 +81,7 @@ def test_cutmix_algorithm(self, fake_data, alpha, uniform_sampling, minimal_stat
 
         algorithm = CutMix(alpha=alpha, uniform_sampling=uniform_sampling, interpolate_loss=interpolate_loss)
         state = minimal_state
-        state.model = ComposerClassifier(torch.nn.Flatten())
+        state.model = ComposerClassifier(torch.nn.Flatten(), num_classes=y_fake.shape[0])
```
Comment (dakinggg) [inline tests/algorithms/test_cutmix.py:84]: Is this right? above, `y_fake = torch.randint(num_classes, size=(N,))`, so I think `y_fake.shape[0]` is the size of the dataset.

File: tests/algorithms/test_mixup.py [hunk 12]
```diff
@@ -57,7 +57,7 @@ def test_mixup_algorithm(self, fake_data, alpha, interpolate_loss, minimal_state
 
         algorithm = MixUp(alpha=alpha, interpolate_loss=interpolate_loss)
         state = minimal_state
-        state.model = ComposerClassifier(torch.nn.Flatten())
+        state.model = ComposerClassifier(torch.nn.Flatten(), num_classes=y_fake.shape[0])
```
Comment (dakinggg) [inline tests/algorithms/test_mixup.py:60]: same

File: tests/callbacks/test_early_stopper.py [hunk 13]
```diff
@@ -27,10 +27,18 @@ def test_early_stopper(metric_sequence: List[float], unit: TimeUnit, device: str
...
+    early_stopper = EarlyStopper('MulticlassAccuracy', dataloader_label, patience=Time(3, unit))
+
+    test_metric_setter = MetricSetterCallback('MulticlassAccuracy',
+                                              dataloader_label,
+                                              Accuracy,
+                                              metric_sequence,
+                                              unit,
+                                              test_device,
+                                              metric_args={
+                                                  'task': 'multiclass',
```
Comment (dakinggg) [inline tests/callbacks/test_early_stopper.py:39]: don't need task here?

File: tests/callbacks/test_threshold_stopper.py [hunk 14]
```diff
@@ -30,10 +30,22 @@ def test_threshold_stopper_eval(metric_sequence: List[float], unit: TimeUnit, de
...
-    test_metric_setter = MetricSetterCallback('Accuracy', dataloader_label, Accuracy, metric_sequence, unit,
-                                              test_device)
+    test_metric_setter = MetricSetterCallback('MulticlassAccuracy',
+                                              dataloader_label,
+                                              Accuracy,
+                                              metric_sequence,
+                                              unit,
+                                              test_device,
+                                              metric_args={
+                                                  'task': 'multiclass',
```
Comment (dakinggg) [inline tests/callbacks/test_threshold_stopper.py:46]: don't need task here?

File: tests/common/datasets.py [hunk 15]
```diff
@@ -40,13 +40,22 @@ class RandomClassificationDataset(Dataset):
 
     def __init__(self, shape: Sequence[int] = (1, 1, 1), size: int = 100, num_classes: int = 2):
         self.size = size
-        self.x = torch.randn(size, *shape)
-        self.y = torch.randint(0, num_classes, size=(size,))
+        self.shape = shape
+        self.num_classes = num_classes
+        self.x = None
+        self.y = None
+        print(self.y)
```
Comment (dakinggg) [inline tests/common/datasets.py:47]: extra print

**Review by dakinggg — APPROVED** [review]
LGTM, approving to unblock. Please paste in proof that 0.10 and 0.11 work

**Comment by mvpatel2000** [comment]
V11 Pass
image
image

V10 
image
image

**Review by mvpatel2000 — APPROVED** [review]
LGTM!
